Antrea supports pass-through/policy-only mode

suwang48404 opened the pull request:
Antrea now may be used in chaining with third party primary CNI to provide network policy enforcement.

Design is at https://github.com/suwang48404/antrea/blob/cloud/docs/policy-only.md.

Also added scripts to create aks/eks cluster which may be used for cloud based feature dev, and CI workflow.

suwang48404 commented:
/test-all

suwang48404 commented:
@tnqn @abhiraut @antoninbas @jianjuns 

hi guys, can u please help review this PR? much appreciated.

For some reason, I cannot modify the reviewers field for this PR.

abhiraut commented:
> @tnqn @abhiraut @antoninbas @jianjuns
> 
> hi guys, can u please help review this PR? much appreciated.
> 
> For some reason, I cannot modify the reviewers field for this PR.

Updated reviewers list. Thanks Su.

antoninbas commented on docs/cloud.md:
```diff
@@ -0,0 +1,69 @@
+   see https://learn.hashicorp.com/terraform/getting-started/install.html
+
+## Create an aks cluster via terraform
+create or obtain a azure service principal. And set the following environment variables from the service principal. For instance in .bashrc
```
Create or obtain an azure service principal.

antoninbas replied:
could you improve the formatting of this file by using subtitles (`###` ) for the different steps, on numbered list items, whichever one looks best to you?

jianjuns commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+# Running Antrea In Pass-Through Mode
+
+Antrea supports chaining with other CNI implementations such as EKS, AKS. In this mode, 
+Antrea enforces k8s network policies, and delegates Pod IP management
```
k8s -> Kubernetes or K8s (the former is better if this is assumed to be user documentation).
NetworkPolicies - capitalize the first letter.

suwang48404 replied:
will do

jianjuns commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+and network connectivity to the primary CNI.
+
+The pass-through mode is designed to be compatible with all major CNI implementations, 
+with emphasis to support cloud provider managed clusters such asAKS, EKS, or potentially
```
suchas -> such as

antoninbas replied:
s/asAKS/as AKS/

antoninbas commented on hack/eks:
```diff
@@ -0,0 +1,131 @@
+# limitations under the License.
+
+# NOTE: terraform enviroment variables need to be set
+INST_PROFILE_NAME=suwang-eks-worker-nodes-1-NodeInstanceProfile-8LL239W4W3GN
```
Please add a check instead to make sure this environment variable is set properly, and print an error message otherwise.

antoninbas commented on hack/eks:
```diff
@@ -0,0 +1,131 @@
+#!/usr/bin/env bash
+
+# Copyright 2019 Antrea Authors
```
For new files, please set Copyright year to 2020.

jianjuns commented on docs/pass-through.md:
````diff
@@ -0,0 +1,237 @@
+
+Create an Antrea manifest for pass-through mode, for instance
+```bash
+./hack/generate-manifest.sh --encap-mode passThrough > ~/my-yaml/antrea-passthrough.yaml
````
I think we should generate this yaml file with make manifest, merge it to repo (like antrea.yml and antrea-ipsec.yml), and publishes it for every release. If different clouds we supported require different yamls, we should generate, merge, and publish those yamls (e.g. antrea-eks.yml, etc.).

suwang48404 replied:
will do.

antoninbas commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+# Running Antrea In Pass-Through Mode
+
+Antrea supports chaining with other CNI implementations such as EKS, AKS. In this mode, 
```
you use `AKS` / `EKS` in this document, while in the other one you used `aks` / `eks`. Could you make it uniform? Maybe use `AKS` / `EKS` everywhere.

antoninbas commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+
+
+### Caveats
+When install Antrea after a k8s cluster already has some Pods running, these pre-installed Pods
```
s/When install/When installing/

antoninbas replied:
suggestion for this whole paragraph:

When installing Antrea on a K8s cluster already running some Pods, Antrea will not be able to enforce network policies on these existing Pods. If Pods are restarted, then Antrea can enforce network policies on them. For instance, to restart all Pods in the coreDNS deployment, you can do the following:

jianjuns commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+These flows should satisfy all Pod traffic patterns with exception of Pod-to-Service traffic
+that we will discuss later.
+
+### Routed CNI Topology
```
Missed a #: ####

jianjuns commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+
+## Design
+
+### Connecting to CNI Topology
```
Connecting -> Inserting ?

jianjuns commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+The above diagram depicts a switched CNI network topology on the left, and what it looks like 
+after Antrea via OVS bridge is inserted into the data path.
+
+In a typical switched CNI topology such as azure AKS, each Pod is connected to a bridge via an veth
```
Since we plan to discuss with AKS team on the integration, I suggest not to describe the current implementation for AKS before the design is closed.
Actually I am not certain if we should say we generally support "switched CNI" or not. Feel we might not really support all "switched CNI".

If we assume "passThrough" is only for cloud. It might make sense to describe the Node/Pod network for each cloud one by one? Not sure if readers without enough contexts understand our terms like switched or routed topology, passthrough, etc.

antoninbas replied:
I had the same thought as Jianjun actually (second paragraph). It may be better to stick to each cloud rather than try to be generic.

suwang48404 replied:
From open source perspective, my goal to expose Antrea to as many CNI as possible, so it is not Cloud specific. It is just so that cloud is big use case right now.

But I'd like to hear what are the draws backs to make solution generic.

jianjuns replied:
I think specific code could be generic, but user documentation and yaml might not be, unless the implementation can be really apply to multiple clouds/platforms without any changes. Let us discuss to understand better.

antoninbas commented on Makefile:
```diff
@@ -125,6 +125,11 @@ tidy:
 
+
+.PHONY: install-cloud-tool
+install-cloud-tool:
```
I'd rather not have this target TBH. What purpose is it serving? There is no reason for `${HOME}/terraform` to be in the path. Since these scripts are likely to be for developers only, I'd rather have them use the `hack/aks` path.

suwang48404 replied:
sound fine. will do

antoninbas commented on docs/cloud.md:
````diff
@@ -0,0 +1,69 @@
+interact with aks cluster
+```
+aks kubectl ... // issue kubectl commands to aks cluster
+aks load ... // load local built images to aks cluster
````
maybe you should indicate that this is mostly useful to load a locally-built version of Antrea into each Node in the cluster.

suwang48404 replied:
will do

antoninbas commented on docs/cloud.md:
````diff
@@ -0,0 +1,69 @@
+```
+interact with aks cluster
+```
+aks kubectl ... // issue kubectl commands to aks cluster
````
I feel like I'm missing something here. Why do I no longer have to use the full `~/terraform/aks` path here? As indicated above, I think we should not install these scripts to `~/terraform/aks`.

suwang48404 replied:
sure I can added the full path.

antoninbas commented on hack/aks:
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
```
What about something with "terraform" in the name, since there is a dependency on it

What about `tf-aks` or `aks-tf` to keep it short?

suwang48404 replied:
will do.

antoninbas commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -136,7 +140,7 @@ func configureContainerAddr(netns ns.NetNS, containerInterface *current.Interfac
 		// ensures the processing of CNI ADD request is not blocked.
 		go func() {
-			netns.Do(func(containerNs ns.NetNS) error {
+			_ = netns.Do(func(containerNs ns.NetNS) error {
```
can we avoid doing this? I don't think it improves style and that's not flagged by any of the linters we run in CI.

suwang48404 replied:
to me, I'd like to think if a method return an error, it needs to be handled, even though it is don't care in this particular use case. It tell user in the context of usage that error does not happen. 

I could not find a good/authentic source to back my reasoning, hope this provide some similarly thought process https://travix.io/unhandled-errors-in-go-3f341f2704dd; but feel free to prove I am wrong.

tnqn replied:
I have the same feeling with @antoninbas, `_ = XXX` and code like `defer func() { _ = f.Close() }()` looks not helpful and a little overkill, especially I have never seen the latter in other projects. Ignoring errors is quite common even in go src code, for example: 
https://github.com/golang/go/blob/cc6a8bd0d7f782c31e1a35793b4e1253c6716ad5/src/net/http/httputil/dump.go#L115
https://github.com/golang/go/blob/cc6a8bd0d7f782c31e1a35793b4e1253c6716ad5/src/net/http/httputil/dump.go#L137

tnqn replied:
Even code in effective go introducing `defer` doesn't do it: https://golang.org/doc/effective_go.html#defer.

antoninbas replied:
If you use the Goland IDE, you can disable the "Unhandled Error" warnings. Unfortunately there doesn't seem to be a way to only disable the warning for `defer` statements.

antoninbas commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+The bridge typically has up-link access to worker Node external, but it is not required for Antrea 
+to work. Intra-Node Pod traffic are switched within the bridge, and inter-Node Pod traffic are 
+switched by bridge via up-link to/from outsider the worker Node. When Pod traffic requires iptables
+ help 
```
that's some weird text wrapping here :/

antoninbas commented on docs/pass-through.md:
````diff
@@ -0,0 +1,237 @@
+sysctl net.bridge.bridge-nf-call-iptables=1
+```
+
+When a Pod is instantiated, the containerd runtime first calls the primary CNI to configure
````
did you actually mean "containerd runtime" here or any "container runtime"?

suwang48404 replied:
I mean containerD runtime as in https://containerd.io.

antoninbas replied:
I know containerd, but why do you use this specific one here? Is this the default one in AKS?

suwang48404 replied:
I know AWS/EKS uses docker, therefore induces that containerD is the backed.

antoninbas commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+In a typical switched CNI topology such as azure AKS, each Pod is connected to a bridge via an veth
+-pair. 
+The bridge typically has up-link access to worker Node external, but it is not required for Antrea 
+to work. Intra-Node Pod traffic are switched within the bridge, and inter-Node Pod traffic are 
```
I believe traffic is singular.

suwang48404 replied:
will do

jianjuns commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+# Running Antrea In Pass-Through Mode
```
passthrough should be one word - so not pass-through, passThrough.
But I guess I need to understand the implementation better, to see if we should still call it passthrough.

antoninbas commented on docs/pass-through.md:
```diff
@@ -0,0 +1,237 @@
+how primary CNI network are connected, it simply needs to program the following OVS flows in 
+the OVS bridge:
+1) ARPs are free flowing as if they are in a regular L2 switch
+2) IP packets with destination MAC matches to an local Pod are switched to that local Pod
```
grammar is wrong

suwang48404 replied:
will check

antoninbas commented on build/images/Dockerfile.ubuntu:
```diff
@@ -2,16 +2,14 @@ FROM ubuntu:18.04 as cni-binaries
 
 FROM antrea/openvswitch:2.13.0
 
+RUN apt-get update && apt-get install -y jq
```
Yes, I also left a comment for this

suwang48404 replied:
will do.

suwang48404 commented on build/images/scripts/install_cni_passthrough:
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
```
will do

suwang48404 commented on build/images/Dockerfile.ubuntu:
```diff
@@ -2,16 +2,14 @@ FROM ubuntu:18.04 as cni-binaries
     apt-get install -y --no-install-recommends wget ca-certificates
-
-# Leading dot is required for the tar command below
-ENV CNI_PLUGINS="./host-local ./loopback"
```
Will check on this.

suwang48404 commented on cmd/antrea-agent/agent.go:
```diff
@@ -159,7 +161,7 @@ func run(o *Options) error {
 		o.config.OVSBridge,
 		nodeConfig.Name,
-		nodeConfig.PodCIDR.String(),
+		fmt.Sprintf("%s", nodeConfig.PodCIDR),
```
so that nodeConfig.PodCIDR may be nil and not crash

suwang48404 replied:
with policy-only mode, podCIDR is not required, as in cloud k8s cluster, podCIDR is not set in NodeSpec

suwang48404 commented:
> @suwang48404 : thanks for the work.
> Sorry, I did not finish review of the PR, but mostly looked at yamls and pass-through.md.
> 
> Three general comments:
> 
> 1. I feel it would be much helpful if you can explain the difference between no-Encap, and passthrough. And we might revisit if we should keep both of them.

===> in nutshell, in noEncap PodCIDR is allocated by Antrea, it therefore controls Pods's IPAM as well as network connectivities; In pass-through mode, podCDIR is allocated by primary CNI, antrea does not know IPAM allocation scheme, and how Pod are connected by primary CNI, and constraints of underlay network.

> 2. I feel CNI chaining (mainly about CNI config file changes?), no-IPAM (IPAM by other CNI), and passthrough should be three different behaviors, even they are all required for cloud. And in code we should not bind them together.
===> we will discuss??

> 3. For AKS, I would suggest to discuss with AKS team first before merging AKS specific code.
> 
====> design is not AKS or EKS, it mean to be generic. open to any CNIs.

> And a question - wonder why you chose Terraform but not other technologies (like Ansible) for AKS/EKS deployment?
======> it just simpler to me to use. It is intention based, a lot of resource on the web to get help and tools

jianjuns commented:
> > 1. For AKS, I would suggest to discuss with AKS team first before merging AKS specific code.
> 
> ====> design is not AKS or EKS, it mean to be generic. open to any CNIs.
 
I mainly meant the documentation and Terraform part. And I still need to understand if the implementation can really be applied to multiple CNIs without changes.
If we want to document the generic implementation, maybe one way is to treat it as design/architecture doc, but create a separate user doc ideally with details for each supported/validated clouds.

suwang48404 commented:
/test-all

suwang48404 commented:
@jianjuns @tnqn @abhiraut @antoninbas 
per request, removed l2 CNI policy only support.

Changes in detals:
1. removed all references to l2 CNI support.
2. removed all logic and support for auto detect CNI topology.
3. removed all test cases pertaining to l2 CNI and auto CNI detection.
4. renamed traffic mode from passThrough to networkPolicyOnly.

jianjuns commented:
Thanks @suwang48404 for the adjustment!

tnqn commented on pkg/agent/iptables/iptables.go:
```diff
@@ -125,10 +125,16 @@ func (c *Client) Initialize(nodeConfig *config.NodeConfig) error {
+			// Append ANTREA-POSTROUTING chain which contains Antrea related postrouting rules to POSTROUTING chain.
+			rule{NATTable, PostRoutingChain, nil, AntreaPostRoutingChain, nil, "Antrea: jump to Antrea postrouting rules"},
+			// Masquerade traffic requiring SNAT (has masqueradeMark set).
+			rule{NATTable, AntreaPostRoutingChain, []string{"-m", "mark", "--mark", masqueradeMark}, MasqueradeTarget, nil, "Antrea: masquerade traffic requiring SNAT"},
```
I have PR #462, one of whose purpose is solving the issue of cleanup of iptables when switching mode or upgrading, you may take a look and rebase on it, hoping it will get merged soon.

suwang48404 replied:
sure.

tnqn commented on pkg/agent/route/route.go:
```diff
@@ -38,8 +39,19 @@ const (
+	SetupPolicyOnlyMode() error
+	MigrateRoutesToGw(linkName string) error
+	UnMigrateRoutesFromGw(route *net.IPNet, linkName string) error
+}
```
I have PR #462 doing similar change but hide more details from the interface's consumer to make windows implementation easier, can you take a look and see if it suits your PR too?

suwang48404 replied:
will merge when PR#4762 is checked in

tnqn commented on pkg/agent/route/route.go:
```diff
@@ -310,10 +322,120 @@ func (c *Client) RemoveServiceRouting() error {
+	return nil
+}
+
+// UnMigrateRoutesToGw move route from gw to link linkName if provided; otherwise route is deleted
```
```suggestion
// UnMigrateRoutesToGw moves route from gw to link linkName if provided; otherwise route is deleted
```

suwang48404 replied:
done

tnqn commented on pkg/agent/route/route.go:
```diff
@@ -310,10 +322,120 @@ func (c *Client) RemoveServiceRouting() error {
+		}
+	}
+
+	// swap address if any
```
Is there any cloud having addresses on its host side peer devices?

suwang48404 replied:
some switched CNI solution assigns IP to PtP device in host network, just take care of that case. In term of cloud provider, EKS does not assign host network IP to PtP device.

jianjuns replied:
Nit: Swap address if any.

tnqn commented on pkg/agent/route/route.go:
```diff
@@ -310,10 +322,120 @@ func (c *Client) RemoveServiceRouting() error {
+}
+
+// UnMigrateRoutesToGw move route from gw to link linkName if provided; otherwise route is deleted
+func (c *client) UnMigrateRoutesFromGw(route *net.IPNet, linkName string) error {
```
method name doesn't match comment

suwang48404 replied:
done

tnqn commented on pkg/agent/util/net_linux.go:
```diff
@@ -0,0 +1,123 @@
+		return nil, "", fmt.Errorf("failed to get link for idx %d: %w", peerIdx, err)
+	}
+
+	// not attached to an bridge.
```
```suggestion
	// not attached to a bridge.
```

suwang48404 replied:
done

tnqn commented on docs/policy-only.md:
```diff
@@ -0,0 +1,193 @@
+except that conntrack takes place not within OVS conntrack zone, 
+but with host network's default conntrack zone.
+
+It have some effect on statistics collection. If original reply traffic reaches 
```
```suggestion
It has some effect on statistics collection. If original reply traffic reaches 
```

suwang48404 replied:
done

tnqn commented on docs/policy-only.md:
```diff
@@ -0,0 +1,193 @@
+3) if DNATed destination in a) is inter-Node Pod or external network entity, is sent back
+to host network space yet again to be forwarded via host network.
+
+We refers the last traffic pattern as re-entrance traffic because in this pattern, a IP packet 
```
```suggestion
We refer the last traffic pattern as re-entrance traffic because in this pattern, an IP packet 
```

suwang48404 replied:
done.

suwang48404 commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -666,3 +631,129 @@ func (pc *podConfigurator) reconcile(pods []corev1.Pod) error {
+
+	containerConfig.OVSPortConfig = &interfacestore.OVSPortConfig{PortUUID: portUUID, OFPort: ofPort}
+	// Add containerConfig into local cache
+	pc.ifaceStore.AddInterface(containerConfig)
```
in current implementation, there was no need for caller to process containerConfig.

suwang48404 replied:
let me know if u feel there is a need to return containerConfig even if there is no taker.

suwang48404 replied:
I see, thx will do

suwang48404 commented on pkg/agent/util/net_linux.go:
```diff
@@ -0,0 +1,123 @@
+func GetNetLink(dev string) netlink.Link {
+	link, err := netlink.LinkByName(dev)
+	if err != nil {
+		klog.Fatalf("cannot find dev %s: %w", dev, err)
```
this is convenient func so that caller do not need to handle error, it is expected to succeed all the time, if not, the underlying link created by agent is removed without agent knowing.

suwang48404 commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -297,10 +297,19 @@ func (c *Controller) reconcile() error {
+	if c.networkConfig.TrafficEncapMode.IsNetworkPolicyOnly() {
+		klog.Infof("TrafficEncap mode is %s, skipping NodeRouteController", c.networkConfig.TrafficEncapMode)
+		_ = c.iptablesClient.Reconcile()
+		<-stopCh
```
good catch.

suwang48404 commented on docs/policy-only.md:
```diff
@@ -0,0 +1,193 @@
+after Antrea via OVS bridge is inserted into the data path.
+
+In a routed CNI network topology, such as AWS EKS, each Pod connects to host network via an point-to
+-point like device, such as (but not limited to) an veth-pair . On the host network, on each PtP
```
done

suwang48404 commented on docs/policy-only.md:
```diff
@@ -0,0 +1,193 @@
+after Antrea via OVS bridge is inserted into the data path.
+
+In a routed CNI network topology, such as AWS EKS, each Pod connects to host network via an point-to
+-point like device, such as (but not limited to) an veth-pair . On the host network, on each PtP
```
done.

suwang48404 commented on docs/policy-only.md:
```diff
@@ -0,0 +1,193 @@
+There are no L2 requirements as all MACS stay within the OVS. 
+Antrea requires no knowledge of Pod network configurations nor
+how CNI network are connected, it simply needs to program the following OVS flows in the OVS bridge:
+1) An default ARP responder flow that answers any ARP request with the same MAC. Its sole
```
done

suwang48404 commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -350,43 +354,20 @@ func (pc *podConfigurator) configureInterfaces(
-	if err != nil {
-		return fmt.Errorf("failed to add Openflow entries for container %s: %v", containerID, err)
+	if err := pc.connectInterfaceToOVS(podName, podNameSpace, containerID, hostIface, containerIface, result.IPs); err != nil {
+		return fmt.Errorf("failed to connect to ovs for contaiiner %s: %v", containerID, err)
```
done

suwang48404 commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -666,3 +631,129 @@ func (pc *podConfigurator) reconcile(pods []corev1.Pod) error {
 	return nil
 }
+
+// connectInterfaces connect an existing interface to ovs br-int.
```
done.

suwang48404 commented on docs/policy-only.md:
```diff
@@ -0,0 +1,193 @@
+When a Pod is instantiated, the container runtime first calls the primary CNI to configure
+Pod's IP and route table, DNS etc, and connects Pod to host network with a Point-To-Point
+device such as a veth-pair.
+Due to chaining, container runtime then calls Antrea, and the Antrea agent attaches Pod's PtoP
```
done. Changes to shorthand PtP.

suwang48404 commented on docs/policy-only.md:
```diff
@@ -0,0 +1,193 @@
+and to be steered back OVS bridge for network policy enforcements.
+
+#### Conntrack
+Note also that the re-entrance traffic, an request for service, after load balancing, routed to
```
done

suwang48404 commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -666,3 +631,129 @@ func (pc *podConfigurator) reconcile(pods []corev1.Pod) error {
+	if err != nil {
+		return fmt.Errorf("failed to open netns %s: %v", containerNetNS, err)
+	}
+	defer func() { _ = netNS.Close() }()
```
done

suwang48404 commented:
@tnqn thx for review. comments addressed.

jianjuns commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -456,9 +502,10 @@ func (c *client) gatewayIPSpoofGuardFlow(gatewayOFPort uint32, category cookie.C
+func (c *client) serviceCIDRDNATFlow(serviceCIDR *net.IPNet, gatewayMAC net.HardwareAddr, gatewayOFPort uint32, category cookie.Category) binding.Flow {
 	return c.pipeline[dnatTable].BuildFlow(priorityNormal).MatchProtocol(binding.ProtocolIP).
 		MatchDstIPNet(*serviceCIDR).
+		Action().SetDstMAC(gatewayMAC).
```
Sorry, but why need this action now?

suwang48404 replied:
unlike in encap case where we know default route is pointing to gw0. In policy-only mode, antrea has no idea what Pod's route table looks like, and it is next hop and Mac, Thus this flow used in conjunction with arpResponderStaticFlow steers service traffic to gw0.

jianjuns replied:
Why not return gw0 MAC in arpResponderStaticFlow?

jianjuns replied:
@tnqn @wenyingd: remind me why we do not use the virtual MAC for gw0?

suwang48404 replied:
I think in encap mode, pod default route nh is to gw0, thus gw0 would response with its correct Mac address, thus no need to use virtual MAC

suwang48404 replied:
we can return gw0 MAC in arpResponderStaticFlow.

tnqn replied:
@jianjuns the arp responder is only installed for remote ips to prevent arp broadcast off the node, we didn't have any arp responder for local ips, regardless of pods or gw0, all of them learn the MAC by getting the real ARP reply. I think it's just we didn't see a good reason to do extra work when the traditional way can work.

jianjuns replied:
@tnqn : my question is if we set gw0 to use virtual MAC (it is also good to reduce chances of MAC conflicts), could our flows be simpler (like to avoid rewrite MAC for remote traffic).

I just do not like to have too many MACs - gw0 MAC, GW virtual MAC, and Su's new virtual MAC. Hard to understand which is for what.

But probably this is irrelevant to this PR.

tnqn replied:
@jianjuns There could be some side effect to make gw0 use the same virtual MAC, the same MAC can appear on tunnel port and gw0 port, which can confuse ovs fdb table, not sure whether we rely on it or not right now, but there was forwarding issue because of it. And when gw0 talks to other nodes, the packets it sent will have same src and dst mac, not sure if it could lead to any issue.
I think currently the vmac is not complex, the vmac just represents the other side of tunnel, both from ingress and egress's perspective, all local macs are real and learnt from ARP. If we share vmac between gw0 and remote endpoints, it sounds a little confusing that which this packet targets for from L2 layer.

Back to this new vmac, per today's discussion, since the packets sent back to OVS for policy enforcement lost its original source IP and ofport, there's no way for policy to take effect, is the mac rewriting still needed?

jianjuns commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -666,3 +631,129 @@ func (pc *podConfigurator) reconcile(pods []corev1.Pod) error {
+		klog.V(2).Infof("Did not find the port for container %s in local cache", containerID)
+		return nil
+	}
+	if err := pc.routeClient.UnMigrateRoutesFromGw(&net.IPNet{
```
Is it needed or we can just remove routes?

suwang48404 replied:
UnMigrateRouteFromGw does some error checking, also good to encapsulate.

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -310,10 +322,120 @@ func (c *Client) RemoveServiceRouting() error {
+	return net.ParseMAC("12:34:56:78:9a:bc")
+}
+
+// SetupPolicyOnlyMode configures routing needed by traffic policy-only mode.
```
in policy-only mode?

suwang48404 replied:
done.

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -310,10 +322,120 @@ func (c *Client) RemoveServiceRouting() error {
+	gwLink := util.GetNetLink(c.nodeConfig.GatewayConfig.Name)
+	_, gwIP, _ := net.ParseCIDR(fmt.Sprintf("%s/32", c.nodeConfig.NodeIPAddr.IP.String()))
+	if err := netlink.AddrReplace(gwLink, &netlink.Addr{IPNet: gwIP}); err != nil {
+		klog.Fatalf("Failed to add address %s to gw %s: %v", gwIP, gwLink.Attrs().Name, err)
```
Should we return here?

suwang48404 replied:
fatal would crash the process.

tnqn replied:
I have concern about using fatal in library and utils code(see GetNetLink) too, thinking these callees shouldn't have assumption that they are critical enough to control the lifecycle of the whole application, otherwise the control would be spread all over pkg code.
It would be more clean if only main control path decides it and having library code just returning error.

jianjuns replied:
I would agree we should not use fatal to exit. Let us return error and handle (finally exit) at the upper layer.

suwang48404 replied:
sure will change.

suwang48404 replied:
removed klog.fatal with return fmt.Errorf in route.go
remove klog.fatal in net_linux.go

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -310,10 +322,120 @@ func (c *Client) RemoveServiceRouting() error {
+// SetupPolicyOnlyMode configures routing needed by traffic policy-only mode.
+func (c *client) SetupPolicyOnlyMode() error {
+	gwLink := util.GetNetLink(c.nodeConfig.GatewayConfig.Name)
+	_, gwIP, _ := net.ParseCIDR(fmt.Sprintf("%s/32", c.nodeConfig.NodeIPAddr.IP.String()))
```
Sorry, I did not know we need an IP on gw0. But how could we know this IP is not used (e.g. reserved from underlay router)?

suwang48404 replied:
perhaps already explained in earlier comment. The IP given to gw0 is exactly the same as Node IP. The reason is so that service traffic to a non-pod server can be masqueraded.

jianjuns replied:
Ok, I now see you add NodeIP to gw0 here. Still I do not know it can work - the same IP can be on two interfaces?

And could you add comments with details why we need to do this? It should be described in your design doc as well.

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -310,10 +322,120 @@ func (c *Client) RemoveServiceRouting() error {
+
+	// add default route to service table.
+	_, defaultRt, _ := net.ParseCIDR("0/0")
+	nhIP := net.ParseIP("169.254.253.1")
```
How this IP is decided? We just need a virtual IP not conflicts with others? If so, define constant.

suwang48404 replied:
move to const

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -310,10 +322,120 @@ func (c *Client) RemoveServiceRouting() error {
+func (c *client) resolveDefaultRouteNHMAC() (net.HardwareAddr, error) {
+	// This MAC is relevant to L2 mode only.
+	// TODO, for now uses AKS default MAC.
+	return net.ParseMAC("12:34:56:78:9a:bc")
```
Why we use AKS MAC? Maybe define a random one.
Use a constant.

suwang48404 replied:
in l3 case, no reason, it can be anything. It matter in L2 (AKS case) where only known MAC may be used for next hop.

suwang48404 commented on pkg/agent/agent.go:
```diff
@@ -375,6 +375,14 @@ func (i *Initializer) setupGatewayInterface() error {
 	}
 
+	if i.networkConfig.TrafficEncapMode.IsNetworkPolicyOnly() {
+		// In policy-only mode, node IP is also assigned to local gateway for masquerade.
```
the same IP used by Node for masquerade on uplink is also used on gw0 for masquerade for service traffic whose backend is non-pod IP..

jianjuns replied:
But we do not add the IP to the gw0 interface right? Maybe rephrase the sentence?

suwang48404 replied:
we *do* add the same Node IP to gw0 for masquerade service traffic whose end point is non-Pod IP.

jianjuns replied:
Sorry, I am still confused. Could you point to me where we add Node IP to gw0? How could two interfaces in the default namespace share the same IP? This is not described in policy-only.md either.

suwang48404 commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -350,43 +354,20 @@ func (pc *podConfigurator) configureInterfaces(
+		return fmt.Errorf("failed to connect to ovs for contaiiner %s: %v", containerID, err)
 	}
 	defer func() {
 		if !success {
```
I think the current logic should be fine, let me know if I missed anything

1. if there is any error after "defer", function return that error
2. but before function returns the error, "defer" function is called. No need to check for error?

suwang48404 commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -666,3 +631,129 @@ func (pc *podConfigurator) reconcile(pods []corev1.Pod) error {
+	containerID := containerConfig.ContainerID
+	klog.V(2).Infof("Deleting Openflow entries for container %s", containerID)
+	if err := pc.ofClient.UninstallPodFlows(containerConfig.InterfaceName); err != nil {
+		return fmt.Errorf("failed to delete Openflow entries for container %s: %v", containerID, err)
```
we can ...
 this just extracts the existing implementation and places them under  disconnectInterfaceFromOVS, there is no change to error handling logic.

jianjuns replied:
I know. Do not mean your change. You can choose to ignore this comment, if you think it should be a different PR.

jianjuns replied:
Keep the comment a while to get @tnqn and @weiqiangt 's notice.

tnqn replied:
@jianjuns If it failed to uninstall flow and succeeded to delete OVS port, the issue of reusing ofport could happen when container runtime calls CNI DEL next time. 
Since we expect the error should lead to CNI DEL retry, it doesn't matter that we don't remove the interface OVS port this round and bind lifecycle of flows and port. In worst case restarting agent can clean up stale ports.

jianjuns replied:
Ok. I guess we need a comment here.

suwang48404 commented on pkg/agent/cniserver/server.go:
```diff
@@ -536,6 +613,18 @@ func (s *CNIServer) reconcile() error {
 	return s.podConfigurator.reconcile(pods.Items)
 }
 
+func (s *CNIServer) checkAndConfigurePolicyOnlyMode() error {
```
this going back to l2/l3 mode auto-detection where l2 or l3 CNI topo cannot be detected on init time. 

I have removed all of those auto-detect logic, but leave this here because CNI server knows about Pod configuration, therefore should be the authoritative source to determine what iptable, route table, and ovs flows should behave.

jianjuns replied:
I do not agree - otherwise other initial route, iptables, and flows should be all moved to CNI server. How policy-only is different from other modes in this front? The current CNI server just configures interfaces and does not know anything about encapMode?

suwang48404 replied:
sounds good, will change

suwang48404 commented on pkg/agent/cniserver/server.go:
```diff
@@ -98,7 +99,9 @@ type CNIServer struct {
 	// podUpdates is a channel for notifying Pod updates to other components, i.e NetworkPolicyController.
-	podUpdates chan<- v1beta1.PodReference
+	podUpdates    chan<- v1beta1.PodReference
+	networkConfig *config.NetworkConfig
```
see above,.

but I don't have strong opinion, feel free to reopen it.

suwang48404 replied:
I think either encapMode or chainedCNI is just information. Either way works for me.

jianjuns replied:
I would suggest to pass only a flag like chainedCNI.

suwang48404 replied:
so far the model has been passing config info to each module, and let module decide for itself what do do. Should CNIserver follow the same model? That is, should we consider caller, cmd/agent.go,  more authoritative to determine what mode CNIServer should run?

jianjuns replied:
I think it is still each module decides what it needs and agent.go just populates what are needed for a module. If you look at the current code, NetworkConfig and EncapMode is passed to only a few modules, but not many others. All modules have different init parameters.

suwang48404 replied:
ok, sounds good. will do.

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -245,11 +253,16 @@ func (c *client) InstallGatewayFlows(gatewayAddr net.IP, gatewayMAC net.Hardware
 	}
 
+	if c.encapMode.SupportsNoEncap() {
+		flows = append(flows, c.reEntranceBypassCTFlow(gatewayOFPort, gatewayOFPort, cookie.Default))
```
it meant to all service traffic including noEncap

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -284,16 +297,17 @@ func (c *client) initialize() error {
 
 	if c.encapMode.SupportsNoEncap() {
-		if err := c.flowOperations.Add(c.l2ForwardOutputInPortFlow(config.HostGatewayOFPort, cookie.Default)); err != nil {
+		if err := c.flowOperations.Add(c.l2ForwardOutputReentInPortFlow(config.HostGatewayOFPort, cookie.Default)); err != nil {
```
it is both noEncap and policy-only. But it is just method rename, no change to substance.

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -284,16 +297,17 @@ func (c *client) initialize() error {
+func (c *client) Initialize(roundInfo types.RoundInfo, nodeConfig *config.NodeConfig, encapMode config.TrafficEncapModeType) (<-chan struct{}, error) {
+	c.nodeConfig = nodeConfig
 	c.encapMode = encapMode
+	c.gatewayPort = config.HostGatewayOFPort // Do better ??
```
move gatewayPort assignment to installGatewayFlows

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -372,3 +386,19 @@ func (c *client) DeleteStaleFlows() error {
 	return c.deleteFlowsByRoundNum(*c.roundInfo.PrevRoundNum)
 }
+
+func (c *client) SetupPolicyOnlyFlows() error {
```
In case that chaining mode cannot be determined by init time.

suwang48404 replied:
called at init time.

suwang48404 commented on docs/policy-only.md:
```diff
@@ -0,0 +1,194 @@
+# Running Antrea In Policy Only Mode
+
+Antrea supports chaining with other CNI implementations such as EKS. In this mode, 
+Antrea enforces kubernetes network policies, and delegates Pod IP management
```
do u consider network policies as k8s kind, therefore I should replace all "network policies"/"network policy" with NetworkPolicy? or network policy as more general term, therefore existing format should be suffice?

jianjuns replied:
I feel in most cases you mean Kuberentes NetworkPolicy, and we do not really have other policies yet. When we do kube-proxy, even for policy-only mode, it is still bye OVS, and so seems it is more NetworkPolicy only?

BTW, %s/kubernetes/Kubernetes, or if you like K8s.

suwang48404 replied:
ok will change
kubernetes -> Kubernetes
network policy/cies -> NetworkPolicy

tnqn commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -297,10 +297,20 @@ func (c *Controller) reconcile() error {
 
 // Run will create defaultWorkers workers (go routines) which will process the Node events from the
-// workqueue.
+// work queue.
```
workqueue is a widely used compound word in Kubernetes, even in linux.
https://github.com/kubernetes/kubernetes/blob/583d83a13557fc4ceb68355fa590c194f23a1721/staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go#L32
https://github.com/torvalds/linux/blob/6f0d349d922ba44e4348a17a78ea51b7135965b1/Documentation/core-api/workqueue.rst#introduction
So maybe no need to separate them.

suwang48404 replied:
done.

suwang48404 commented on build/images/scripts/install_cni_policy_only:
```diff
@@ -0,0 +1,17 @@
+#!/usr/bin/env bash
```
renames install_cni_policy_only to install_cni_channing,
renames build/yamls/patch/policy-only/installCNI.yml to build/yamls/patches/chaining/installCNI.yml

suwang48404 commented on build/yamls/base/conf/antrea-agent.conf:
```diff
@@ -38,4 +38,6 @@
 # noEncap: Inter-node Pod traffic is not encapsulated, but Pod to outbound traffic is masqueraded.
 #          Underlying network must be capable of supporting Pod traffic across IP subnet.
 # hybrid: noEncap if worker Nodes on same subnet, otherwise encap.
+# networkPolicyOnly: Utilizes CNI chaining, the primary CNI is responsible for Pod IPAM and connectivity.
```
done.

suwang48404 commented on cmd/antrea-agent/config.go:
```diff
@@ -70,5 +70,7 @@ type AgentConfig struct {
 	//          Underlying network must be capable of supporting Pod traffic across IP subnet.
 	// Hybrid: noEncap if worker Nodes on same subnet, otherwise encap.
+	// NetworkPolicyOnly: Utilizes CNI chaining, the primary CNI is responsible for Pod IPAM and connectivity.
+	//                   Antrea enforces network policies.
```
done.

suwang48404 commented on docs/assets/l3-cni.svg:
```diff
@@ -0,0 +1,136 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
1. C1-> P1
2. l3-cni.svg -> policy-only.svg

suwang48404 commented on pkg/agent/cniserver/server.go:
```diff
@@ -521,6 +555,49 @@ func (s *CNIServer) Run(stopCh <-chan struct{}) {
 
+// interceptAdd handles Add request in policy only mode. Another CNI must already
+// be called prior to Antrea CNI to allocate IP and ports. Antrea takes allocated port
+// and hooks it to ovs br-int.
```
done

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -218,6 +221,11 @@ func (c *client) InstallPodFlows(containerID string, podInterfaceIP net.IP, podI
 	}
+	if c.encapMode.IsNetworkPolicyOnly() {
+		flows = append(flows,
+			c.l3ToPodFlow(podInterfaceIP, podInterfaceMAC, cookie.Pod),
```
done

suwang48404 commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -322,6 +311,18 @@ func (c *client) l2ForwardOutputInPortFlow(gwPort uint32, category cookie.Catego
 
+// l3BypassMacRewriteFlow bypasses remaining l3forwarding flows if the MAC is set via ctRewriteDstMACFlow in
+// conntrackState stage.
+func (c *client) l3BypassMacRewriteFlow(gatewayMAC net.HardwareAddr, category cookie.Category) binding.Flow {
```
done.

suwang48404 commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -337,6 +338,32 @@ func (c *client) l3FlowsToPod(localGatewayMAC net.HardwareAddr, podInterfaceIP n
+// l3ToGWFlow generates the flow to rewrite MAC to gw port if the packet is received
+// is unmatched by local Pod flows.
+// This flow is used in policy only traffic mode.
+func (c *client) l3ToGWFlow(gwMAC net.HardwareAddr, category cookie.Category) binding.Flow {
```
done.

suwang48404 commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -408,6 +435,25 @@ func (c *client) arpResponderFlow(peerGatewayIP net.IP, category cookie.Category
 }
 
+// arpResponderStaticFlow generates ARP request for any IP with the same global IP.
+// This flow is used in l3 policy only, where traffic are routed via IP not MAC.
```
done.

suwang48404 commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -408,6 +435,25 @@ func (c *client) arpResponderFlow(peerGatewayIP net.IP, category cookie.Category
 		Done()
 }
 
+// arpResponderStaticFlow generates ARP request for any IP with the same global IP.
```
done.

suwang48404 commented on docs/policy-only.md:
```diff
@@ -0,0 +1,194 @@
+## Design
+
+### Inserting to CNI Topology
+Antrea is designed to work as network policy plug-in for the following CNI topology models. For
```
done.

suwang48404 commented on docs/policy-only.md:
```diff
@@ -0,0 +1,194 @@
+as long as a CNI implementation fits into these models, Antrea may be inserted to enforce
+network policy in that CNI's environment.
+
+### Routed CNI Topology
```
done.

suwang48404 commented:
@jianjuns Thx for review. I should addressed all ur comments. As there are many inputs, please let me know if I missed any.
thx, Su

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -372,3 +386,19 @@ func (c *client) DeleteStaleFlows() error {
 }
+
+func (c *client) SetupPolicyOnlyFlows() error {
+	if err := c.flowOperations.Add(c.l3BypassMacRewriteFlow(c.nodeConfig.GatewayConfig.MAC, cookie.Default)); err != nil {
```
done

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -284,16 +295,23 @@ func (c *client) initialize() error {
 			return fmt.Errorf("failed to install L2 forward same in-port and out-port flow: %v", err)
 		}
 	}
+	if c.encapMode.IsNetworkPolicyOnly() {
```
yes. flow priority should have been taken care of, ideally. Please let me know if I missed anything.

suwang48404 replied:
In policy-only mode, there is no per node podCDIR, therefore inter-node flows do not exist.

suwang48404 commented on pkg/agent/agent.go:
```diff
@@ -256,17 +256,18 @@ func persistRoundNum(num uint64, bridgeClient ovsconfig.OVSBridgeClient, interva
 // time.
 func (i *Initializer) initOpenFlowPipeline() error {
 	roundInfo := getRoundInfo(i.ovsBridgeClient)
+	gateway, _ := i.ifaceStore.GetInterface(i.hostGateway)
```
I believe ifaceStore.GetInterface is retrieving from memory cache not ovsdb. nonetheless, added check for "ok"

suwang48404 commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -408,6 +435,25 @@ func (c *client) arpResponderFlow(peerGatewayIP net.IP, category cookie.Category
+// arpResponderStaticFlow generates ARP reply for any ARP request with the same global virtual MAC.
+// This flow is used in policy-only mode, where traffic are routed via IP not MAC.
+func (c *client) arpResponderStaticFlow(category cookie.Category) binding.Flow {
+	return c.pipeline[arpResponderTable].BuildFlow(priorityNormal).MatchProtocol(binding.ProtocolARP).
```
It should be OK, arpResponderStaticFlow is the *only* flow in arpResponder table in policy-only mode.

suwang48404 commented on pkg/agent/openflow/client.go:
```diff
@@ -372,3 +390,19 @@ func (c *client) DeleteStaleFlows() error {
+
+func (c *client) setupPolicyOnlyFlows() error {
+	// Bypasses remaining l3forwarding flows if the MAC is set via ctRewriteDstMACFlow.
+	if err := c.flowOperations.Add(c.l3BypassMACRewriteFlow(c.nodeConfig.GatewayConfig.MAC, cookie.Default)); err != nil {
```
There seems two ways in adding flows in client.go
I follows the model in initialize(), let me know u'd like me to change still.

jianjuns replied:
AddAll() is always better - it adds/updates flows in a single bundle/transaction. Guess you need to change InstallGatewayFlows() for this.

suwang48404 replied:
sure will do.

suwang48404 replied:
done

suwang48404 commented:
There is merge conflict, will do an final merge and validation once all reviewers are happy with the PR.

jianjuns commented on pkg/agent/openflow/client.go:
```diff
@@ -218,6 +218,12 @@ func (c *client) InstallPodFlows(containerID string, podInterfaceIP net.IP, podI
 		flows = append(flows, c.l3FlowsToPod(gatewayMAC, podInterfaceIP, podInterfaceMAC, cookie.Pod))
 	}
+	if c.encapMode.IsNetworkPolicyOnly() {
+		// in policy-only mode, traffic to local Pod is routed based on destination IP.
```
Nit: in -> In

suwang48404 replied:
done

jianjuns commented on pkg/agent/openflow/client.go:
```diff
@@ -372,3 +390,19 @@ func (c *client) DeleteStaleFlows() error {
+	if err := c.flowOperations.Add(c.l3BypassMACRewriteFlow(c.nodeConfig.GatewayConfig.MAC, cookie.Default)); err != nil {
+		return fmt.Errorf("failed to install l3BypassMACRewriteFlow: %v", err)
+	}
+	// Rewrites MAC to gw port if the packet is received is unmatched by local Pod flows.
```
the packet received is...?

suwang48404 replied:
done

jianjuns commented on pkg/agent/route/route.go:
```diff
@@ -36,10 +37,22 @@ const (
 	routeTableConfigPath  = "/etc/iproute2/rt_tables"
 	// AntreaIPRulePriority is Antrea IP rule priority
 	AntreaIPRulePriority = 300
+	defaultGWIP          = "169.254.253.1"
```
Let us use another name to differentiate with default GW (gw0), like virtualGWIP/MAC.

Please add some comments to explain them, and mention they are for policy-only mode.

suwang48404 replied:
done.

jianjuns commented:
@suwang48404 : actually I also suggest to move policy-only.md to a separate PR - it is long, and might take some time to finish review.

abhiraut commented on pkg/agent/cniserver/server.go:
```diff
@@ -521,6 +552,49 @@ func (s *CNIServer) Run(stopCh <-chan struct{}) {
+	klog.Infof("CNI Chaining: add")
+	prevResult, response := s.parsePrevResultFromRequest(cniConfig.NetworkConfig)
+	if response != nil {
+		klog.Infof("Failed to parse prev result for container %s.", cniConfig.ContainerId)
```
nit: remove period from the end of msg in Infof

suwang48404 replied:
done

suwang48404 commented:
per JianJun's request. also removed policy-only.md associated assert policy-only-cni.svg

abhiraut commented on pkg/agent/cniserver/server.go:
```diff
@@ -521,6 +552,49 @@ func (s *CNIServer) Run(stopCh <-chan struct{}) {
+}
+
+func (s *CNIServer) interceptDel(cniConfig *CNIConfig) (*cnipb.CniCmdResponse, error) {
+	// wait for patch port to be configured
```
old comment?

suwang48404 replied:
removed

abhiraut commented on pkg/agent/cniserver/server.go:
```diff
@@ -521,6 +552,49 @@ func (s *CNIServer) Run(stopCh <-chan struct{}) {
+
+func (s *CNIServer) interceptCheck(_ *CNIConfig) (*cnipb.CniCmdResponse, error) {
+	klog.Infof("CNI Chaining: check")
+	// TODO, check for host interface setup later
```
will this be in a follow up PR? or this PR itself

suwang48404 replied:
perhaps in foreseeable future, this only make sense if Antrea wants to do some additional checking ...

abhiraut commented on pkg/agent/controller/noderoute/node_route_controller.go:
```diff
@@ -301,6 +301,16 @@ func (c *Controller) reconcile() error {
+	// underlying network. Therefore it needs not know the routes to
+	// peer Pod CIDRs.
+	if c.networkConfig.TrafficEncapMode.IsNetworkPolicyOnly() {
+		klog.Infof("TrafficEncap mode is %s, skipping NodeRouteController", c.networkConfig.TrafficEncapMode)
```
do you have to do a c.networkConfig.TrafficEncapMode.String() ?

suwang48404 replied:
I think it should be OK, as "%s" formatter takes String method of variable TrafficEncapMode.

suwang48404 commented on docs/cloud.md:
```diff
@@ -0,0 +1,69 @@
+# Deploying Antrea on a cloud provider
```
weird,. cloud.md is not in this PR anymore. It is now on https://github.com/vmware-tanzu/antrea/pull/466.

suwang48404 replied:
cloud.md moved to https://github.com/vmware-tanzu/antrea/pull/466

suwang48404 commented on docs/cloud.md:
````diff
@@ -0,0 +1,69 @@
+```bash
+~/terraform/aks
+```
+interact with aks cluster
````
same

suwang48404 replied:
cloud.md moved to https://github.com/vmware-tanzu/antrea/pull/466

suwang48404 commented on docs/pass-through.md:
````diff
@@ -0,0 +1,237 @@
+```bash
+./ci/run-k8s-e2e-tests.sh --e2e-all
+```
+You may need to install a lower version of sonobuoy 
````
pass-through.md is no longer part of this PR.

suwang48404 commented on hack/aks:
```diff
@@ -0,0 +1,113 @@
+#!/usr/bin/env bash
+
+# Copyright 2019 Antrea Authors
```
aks is moved to https://github.com/vmware-tanzu/antrea/pull/466

suwang48404 commented on build/yamls/patches/chaining/installCni.yml:
```diff
@@ -0,0 +1,15 @@
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: antrea-agent
```
I think labels are at daemonSet and Pod level and that should already be handled by Kustomize? This only tells Kustomize that what command to use in container InitContainer.

suwang48404 commented on docs/policy-only.md:
````diff
@@ -0,0 +1,194 @@
+
+Create an Antrea manifest for policy-only mode, for instance
+```bash
+./hack/generate-manifest.sh --encap-mode networkPolicyOnly > ~/my-yaml/antrea-policy.yaml
````
we can do that, but user would still need to modify the checked in manifest, because in each CNI environment, service CIDR, MTU may all be different, So it sort defeats the reasoning to have a check-in version of manifest in the first place.

suwang48404 commented on docs/policy-only.md:
````diff
@@ -0,0 +1,194 @@
+```
+You may need to install a lower version of sonobuoy 
+[here](https://github.com/vmware-tanzu/sonobuoy/releases/), and set environment variable SONOBUOY to
+the path to the executable, if the Kubernetes cluster is lower version;
````
policy-only.md is no longer part of this PR.

suwang48404 commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -666,3 +630,130 @@ func (pc *podConfigurator) reconcile(pods []corev1.Pod) error {
+		return nil, fmt.Errorf("failed to add Openflow entries for container %s: %v", containerID, err)
+	}
+	defer func() {
+		if err != nil {
```
u r correct we don't need the defer function at the moment. will remove

suwang48404 commented on pkg/agent/cniserver/pod_configuration.go:
```diff
@@ -666,3 +630,130 @@ func (pc *podConfigurator) reconcile(pods []corev1.Pod) error {
+func (pc *podConfigurator) disconnectInterceptedInterface(podName, podNamespace, containerID string) error {
+	containerConfig, found := pc.ifaceStore.GetContainerInterface(podName, podNamespace)
+	if !found {
+		klog.V(2).Infof("Did not find the port for container %s in local cache", containerID)
```
the log level is copied from removeInterfaces when cannot find interface, I can definitely change this ..

suwang48404 commented:
@tnqn @jianjuns @wenyingd @abhiraut @antoninbas 

Hi guys, I am hoping to get this merged before 0.5, which is early next week. I'd appreciate if u let me know if there are remaining issues with this PR u'd like me to address

thx

antoninbas commented:
@suwang48404 0.5 is in 2 weeks, we will merge this in time

jianjuns commented on pkg/agent/route/route_linux.go:
```diff
@@ -567,3 +598,110 @@ func disableICMPSendRedirects(intfName string) error {
+		return fmt.Errorf("failed to add address %s to gw %s: %v", gwIP, gwLink.Attrs().Name, err)
+	}
+
+	// add default route to service table.
```
Nit: Add

suwang48404 replied:
done.

jianjuns commented on pkg/agent/route/route_linux.go:
```diff
@@ -567,3 +598,110 @@ func disableICMPSendRedirects(intfName string) error {
+	if err := netlink.RouteReplace(route); err != nil {
+		return fmt.Errorf("failed to add default route to service table: %v", err)
+	}
+	// add static neighbor to next hop so that no ARPING is ever required on gw0
```
Add
"." after "gw0".

suwang48404 replied:
done.

jianjuns commented on pkg/agent/route/route_linux.go:
```diff
@@ -567,3 +598,110 @@ func disableICMPSendRedirects(intfName string) error {
+		}
+	}
+
+	// swap address if any
```
Ditto

suwang48404 replied:
done.

jianjuns commented:
@tnqn @wenyingd : do you have other major comments on this PR? Esp. please share if you have better ideas on ClusterIP Services with external endpoints.

tnqn commented on pkg/agent/agent.go:
```diff
@@ -345,6 +349,14 @@ func (i *Initializer) setupGatewayInterface() error {
 
+	if i.networkConfig.TrafficEncapMode.IsNetworkPolicyOnly() {
+		// In policy-only mode, Node IP is also assigned to local gateway for masquerade.
+		i.nodeConfig.GatewayConfig = &config.GatewayConfig{Name: i.hostGateway, MAC: link.Attrs().HardwareAddr, IP: i.nodeConfig.NodeIPAddr.IP}
```
@jianjuns @suwang48404 @wenyingd If masquerading here is to send packets back to OVS pipeline for network policy enforcement, this is not going to work as the packets has no original source ip and of_port, network policy flows cannot identify packets need to filter.
Did I miss anything?

suwang48404 replied:
Hi Quan, u r correct. Network Policy does not work for external service in policy-only mode. I plan to file an issue once this is checked in, explaining the problem and potential remedy.

suwang48404 replied:
I believe nodeIP assigning to gw0 is discussed and resolved during our offline chatting?

Resolving.

tnqn commented on pkg/agent/route/route_linux.go:
```diff
@@ -198,6 +208,12 @@ func (c *Client) initIPTables() error {
+			"-A", antreaMangleChain,
+			"-m", "comment", "--comment", `"Antrea: unmark post LB service packets"`,
+			"-i", c.hostGateway, "!", "-d", c.serviceCIDR.String(),
+			"-j", iptables.MarkTarget, "--set-xmark", "0/0xffffffff",
```
should it only reset the bit used for service mark? otherwise other marks by kubernetes will be reset too.

suwang48404 replied:
I think we want to clean *all* marks for re-entrance traffic, because we do not want to have marks from previous entrance (i.e LB) interfere with current (forwarding) entrance.
Also it is set in PREROUTING stage mangle table, any subsequent K8s marking should not be impacted by this.

tnqn replied:
Got it.

suwang48404 commented:
/test-all

suwang48404 commented:
@jianjuns @tnqn @abhiraut @antoninbas 

If there is no other comments, can anyone help approve and merge the PR, as I have no write privilege to Antrea repro.

suwang48404 commented on pkg/agent/openflow/pipeline.go:
```diff
@@ -337,6 +338,31 @@ func (c *client) l3FlowsToPod(localGatewayMAC net.HardwareAddr, podInterfaceIP n
 
+// l3ToPodFromGwFlow generates the flow to rewrite MAC if the packet IP matches an local IP.
+// This flow is used in policy only traffic mode.
+func (c *client) l3ToPodFlow(podInterfaceIP net.IP, podInterfaceMAC net.HardwareAddr, category cookie.Category) binding.Flow {
```
I was not planning to create a separate ovs table for this check in. Perhaps it may be discussed later, if needed, with another PR?

jianjuns commented:
@susanwu88: I will take another look today. A question - is the poliy-only yaml can be applied to EKS cluster directly? If so, I would merge an antrea-eks.yml to the repo, and publish it in every release. We can make a separate PR for that though.

suwang48404 commented:
> @susanwu88: I will take another look today. A question - is the poliy-only yaml can be applied to EKS cluster directly? If so, I would merge an antrea-eks.yml to the repo, and publish it in every release. We can make a separate PR for that though.

@abhiraut asked same questions. Unfortunately no. Because user would need to change MTU and serviceCIDR in Antrea-agent.conf. The detail is described in policy-only.md coming in after this PR.

suwang48404 commented:
per offline discussion with @jianjuns, he wanted
1. a ./build/yamls/antrea-eks.yml for EKS 
2. a doc/eks-installation.md describing antrea deployment to EKS.

suwang48404 commented:
/test-all

jianjuns commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,34 @@
+Assuming you already have an EKS cluster, and have KUBECONFIG environment variable pointing to<br>
+the kubeconfig file of that cluster.
+
+Based on EKS worker Node MTU size and service CIDR, adjust defaultMTU and serviceCIDR values of<br>
```
Better to make it more clear, like "Kubernetes Service Cluster IP range"

jianjuns replied:
Highlight `defaultMTU` and `serviceCIDR`.

jianjuns replied:
Should we mention MTU of EKS VMs is typically 9001? Maybe better to add a yaml path to change the default MTU to 9001 too.

antoninbas replied:
+1 if we know the typical MTU for EKS nodes, then let's use that (minus encap overhead) as the defaultMTU parameter in antrea-eks.yml.

Once this PR is merged, we can also look into automatic MTU detection

suwang48404 replied:
Leave MTU size untouched because aws ec2 mtu is configurable. 

Rest is done.

jianjuns commented on docs/eks-installation.md:
````diff
@@ -0,0 +1,34 @@
+You can validate it either manually or by running NetworkPolicy tests as follows:
+
+```bash
+./ci/run-k8s-e2e-tests.sh --e2e-network-policy
````
This is not for end users. Suggest to remove this part.

Maybe you can refer to getting-started.md on how to check Antrea components are running (there is a gif there).

suwang48404 replied:
done.

jianjuns commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,34 @@
+# Deploying Antrea in AWS EKS
```
We might add a link to getting-started.md to this doc. @antoninbas

antoninbas replied:
sounds good to me. Next to the "Deploying Antrea in Kind" section, we can have a "Deploying Antrea in Cloud Managed K8s Services" section, to indicate that for now we have EKS support and point to this doc.

suwang48404 replied:
may be another PR?

jianjuns commented:
LGTM. Some minor comments on the documentation.

suwang48404 commented:
/test-all

suwang48404 commented:
/test-all

jianjuns commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,26 @@
+the kubeconfig file of that cluster.
+
+Based on EKS worker Node MTU size and Kubernetes service cluster IP range, adjust
+**defaultMTU** and **serviceCIDR** values of antrea-agent.conf in ./build/yamls/antrea-eks.yml
```
We should add antrea-eks.yml to the release files. For now, probably you can use this way:
"[the EKS deployment yaml](/build/yamls/antrea-eks.yml)"

suwang48404 replied:
done. Modified release.md to include antrea-eks.yml in generate-manifest section.

jianjuns commented on docs/eks-installation.md:
````diff
@@ -0,0 +1,26 @@
+./build/yamls/antrea-eks.yml to the EKS cluster.
+
+```bash
+kubectl apply -f ./build/yamls/antrea-eks.yaml 
````
Probably just: kubectl apply -f antrea-eks.yml

Or if you like check the description at: https://github.com/vmware-tanzu/antrea/blob/master/docs/ipsec-tunnel.md

suwang48404 replied:
done. good pointer. added instruction to download antrea-eks.yml.

suwang48404 commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,26 @@
+
+### Caveats
+
+Some Pod may already be installed before Antrea deployment, Antrea cannot enforce NetworkPolicy<br>
```
1. I thought u specifically said that in Antrea when refer to k8s resource Kind, we use singular, no?

2. so essentially, u r suggesting "Some Pod(s) .... cannot enforce NetworkPolicy", does it sound right to u?

suwang48404 replied:
Ok. I think I misunderstand the comments. Break into two sentences,
"Some Pod may already be installed before Antrea deployment.  Antrea cannot enforce NetworkPolicy. .... "

jianjuns replied:
No we do not use singular for resources (when it means resource type yes, but here you wrote "some Pods".

suwang48404 replied:
done

suwang48404 commented on docs/eks-installation.md:
````diff
@@ -0,0 +1,26 @@
+```bash
+kubectl apply -f ./build/yamls/antrea-eks.yaml 
+```
+Now Antrea should be plugged into the EKS CNI and is ready to enforces NetworkPolicy.<br> 
````
I am sorry, No need to add what?

jianjuns replied:
I meant < br >. Sorry.

suwang48404 replied:
done.

suwang48404 commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,26 @@
+
+This document describes steps to deploy Antrea in NetworkPolicy only mode to an AWS EKS cluster.
+
+Assuming you already have an EKS cluster, and have KUBECONFIG environment variable pointing to<br>
```
I cannot be KBUECONFIG, no?

jianjuns replied:
I meant adding `` around KUBECONFIG

suwang48404 replied:
done.

abhiraut commented on build/yamls/antrea-eks.yml:
```diff
@@ -0,0 +1,595 @@
+    # Make sure it doesn't conflict with your existing interfaces.
+    #hostGateway: gw0
+
+    # Encapsulation mode for communication between Pods across Nodes, supported values:
```
do we want to remove some of these options which are not relevant for eks?

suwang48404 replied:
make sense, may be on a separate PR? I think today all build/yamls/antrea-xxx.yml comes from the same build/base/antrea-agent.conf.

suwang48404 commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,26 @@
+### Caveats
+
+Some Pod may already be installed before Antrea deployment, Antrea cannot enforce NetworkPolicy<br>
+on these pre-installed Pod. This may be remedied by restart the Pod, i.e
```
done.

abhiraut commented on docs/eks-installation.md:
````diff
@@ -0,0 +1,30 @@
+```bash
+kubectl apply -f antrea-eks.yaml 
+```
+Now Antrea should be plugged into the EKS CNI and is ready to enforces NetworkPolicy.
````
enforces -> enforce

suwang48404 replied:
done.

abhiraut commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,30 @@
+### Caveats
+
+Some Pod may already be installed before Antrea deployment. Antrea cannot enforce NetworkPolicy
+on these pre-installed Pod. This may be remedied by restart the Pod, i.e.
```
remedied by restart of the Pod, for example:

suwang48404 replied:
done.

suwang48404 commented:
/skip-all

jianjuns commented on docs/eks-installation.md:
````diff
@@ -0,0 +1,30 @@
+To deploy the latest version of Antrea (built from the master branch) to EKS, get the Antrea EKS
+deployment yaml at:
+```
+https://raw.githubusercontent.com/vmware-tanzu/antrea/master/build/yamls/antrea-eks.yml
````
The link does not work for me. Should be: https://github.com/vmware-tanzu/antrea/blob/master/docs/ipsec-tunnel.md?

suwang48404 replied:
we just need to check in build/yamls/build/antrea-eks.yml, right? Do I miss anything? I suspect link does not work because it is not checked in yet. I use the same link u have inhttps://github.com/vmware-tanzu/antrea/blob/master/docs/ipsec-tunnel.md for IPSec i.e

https://raw.githubusercontent.com/vmware-tanzu/antrea/master/build/yamls/antrea-ipsec.yml

jianjuns replied:
Ok. My mistake. Sorry.

jianjuns commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,30 @@
+
+This document describes steps to deploy Antrea in NetworkPolicy only mode to an AWS EKS cluster.
+
+Assuming you already have an EKS cluster, and have KUBECONFIG environment variable pointing to
```
have the `KUBECONFIG` environment variable point to?

suwang48404 replied:
done

jianjuns commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,30 @@
+
+### Caveats
+
+Some Pod may already be installed before Antrea deployment. Antrea cannot enforce NetworkPolicy
```
Pod -> Pods

jianjuns commented on docs/eks-installation.md:
```diff
@@ -0,0 +1,30 @@
+### Caveats
+
+Some Pod may already be installed before Antrea deployment. Antrea cannot enforce NetworkPolicy
+on these pre-installed Pod. This may be remedied by restarting the Pod. For example,
```
Pod -> Pods
the Pod -> the Pods

suwang48404 replied:
done.

suwang48404 commented:
/test-all

suwang48404 commented:
/test-all

suwang48404 commented:
/test-e2e

suwang48404 commented:
/test-e2e
